### PR TITLE
Prepare multi gpu support

### PR DESCRIPTION
**Pull request by agnesLeroy**
closes: _please link all relevant issues_

### PR content/description
This PR prepares the code base for multi-gpu support.
At the moment all functions fall back to using only GPU 0, but we'll be able to add multi-GPU support little by little on top of this.

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by agnesLeroy** [comment]
@tmontaigu I've done what you suggested and created a `MultiGpuFunctionExecutor` for the tests, but actually I'm thinking a better way of doing this would be to pass a &[CudaStream] to all entry points, and they would make sure GPU 0 is only ever used. Otherwise in the code as it is now, users could try to execute things on GPU 1, but that would conflict with the multi-GPU implementation of `add`, which supposes the keys, inputs & outputs are on GPU 0.

**Comment by tmontaigu** [comment]
If you think that is better you can do it

**Comment by mironov788** [comment]
Very interesting!!

**Comment by agnesLeroy** [comment]
@tmontaigu I gave a tour of the changes to @pdroalves earlier, so he has a good view of what's in the PR normally. I can go over it with you as well if you want, as this is a lot of changes...

**Comment by agnesLeroy** [comment]
@tmontaigu I gave a tour of the changes to @pdroalves earlier, so he has a good view of what's in the PR normally. I can go over it with you as well if you want, as this is a lot of changes...

**Comment by agnesLeroy** [comment]
@tmontaigu I gave a tour of the changes to @pdroalves earlier, so he has a good view of what's in the PR normally. I can go over it with you as well if you want, as this is a lot of changes...

File: tfhe/src/integer/gpu/mod.rs [hunk 1]
```diff
@@ -126,2807 +126,1927 @@ where
...
+pub unsafe fn scalar_addition_integer_radix_assign_async<T: UnsignedInteger>(
+    streams: &CudaStreams,
+    lwe_array: &mut CudaVec<T>,
+    scalar_input: &CudaVec<T>,
+    lwe_dimension: LweDimension,
+    num_samples: u32,
+    message_modulus: u32,
+    carry_modulus: u32,
+) {
+    assert_eq!(
```
Comment (agnesLeroy) [inline tfhe/src/integer/gpu/mod.rs:167]: @pdroalves checkes added here for the GPU indexes

File: tfhe/src/high_level_api/global_state.rs [hunk 2]
```diff
@@ -146,22 +146,21 @@ where
...
-    CudaStream::new_unchecked(device)
+fn create_cuda_cell_streams() -> CudaStreams {
+    CudaStreams::new_multi_gpu()
 }
 
 #[cfg(feature = "gpu")]
 thread_local! {
-    static CUDA_STREAM: std::cell::OnceCell<CudaStream> = std::cell::OnceCell::from(create_cuda_cell_stream());
+    static CUDA_STREAMS: std::cell::OnceCell<CudaStreams> = std::cell::OnceCell::from(create_cuda_cell_streams());
 }
```
Comment (tmontaigu) [inline tfhe/src/high_level_api/global_state.rs:156]: We could inline the `create_cuda_cell_streams`

```rust
#[cfg(feature = "gpu")]
thread_local! {
    static CUDA_STREAMS: std::cell::OnceCell<CudaStreams> = std::cell::OnceCell::from(CudaStreams::new_multi_gpu());
}
```

File: tfhe/src/core_crypto/gpu/mod.rs [hunk 3]
```diff
@@ -12,518 +12,511 @@ use crate::core_crypto::prelude::{
 pub use algorithms::*;
 pub use entities::*;
 use std::ffi::c_void;
-use tfhe_cuda_backend::cuda_bind::*;
+pub(crate) use tfhe_cuda_backend::cuda_bind::*;
 
-#[derive(Debug, Copy, Clone)]
-pub struct CudaPtr {
-    ptr: *mut c_void,
-    device: CudaDevice,
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply tfhe/src/core_crypto/gpu/mod.rs:20]: Ah yes, makes sense :)

File: tfhe/src/core_crypto/gpu/mod.rs [hunk 4]
```diff
@@ -12,518 +12,511 @@ use crate::core_crypto::prelude::{
...
-}
-
-impl CudaStream {
-    pub fn new_unchecked(device: CudaDevice) -> Self {
-        let gpu_index = device.gpu_index();
-        unsafe {
-            let ptr = cuda_create_stream(gpu_index);
+impl CudaStreams {
+    pub fn new_multi_gpu() -> Self {
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply tfhe/src/core_crypto/gpu/mod.rs:26]: I wrote comments for the single gpu and multi gpu `new` functions, do you think it's enough? I will add a section in the user documentation at the end of the quarter as well.

**Comment by agnesLeroy** [comment]
@tmontaigu thanks a lot for your feedback! I pushed a separate commit where I took your comments into account.

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**